pkg/sdk/reconciler.go

```diff
@@ -0,0 +1,83 @@
+package sdk
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	ctlr "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+type reconciler[T ObjectConstraint] struct {
+	client.Client
+	impl         Implementation[T]
+	resourceKind string
+}
+
+var _ reconcile.Reconciler = &reconciler[client.Object]{}
+
+func RegisterController[T ObjectConstraint](mgr manager.Manager, impl Implementation[T]) error {
+	return RegisterControllerWithEventFilter(mgr, impl, nil)
+}
+
+// To handle new controllers from PR https://github.com/nginxinc/nginx-kubernetes-gateway/pull/221
+func RegisterControllerWithEventFilter[T ObjectConstraint](
```

Comment:
The simplest approach I see is to get rid of Register function at all.

So in the manager, do something like this (error handling is not shown):

```
mgr.GetFieldIndexer().IndexField(context.TODO(), &discoveryV1.EndpointSlice{}, KubernetesServiceNameIndexField, ServiceNameIndexFunc)
ctlr.NewControllerManagedBy(mgr).
		For(&discoveryV1.EndpointSlice{}).
		WithEventFilter(predicate.GenerationChangedPredicate{}).
		Complete(NewReconciler(...))

ctlr.NewControllerManagedBy(mgr).
		For(&discoveryV1.Gateway{}).
		Complete(NewReconciler(...))

... 
```

> What do you think about making this more extensible by mirroring the builder pattern of the controller?

That could also work:
```
sdk.RegisterControllerManagedBy(mgr).
   For(&discoveryV1.EndpointSlice{}).
   WithEventFilter(predicate.GenerationChangedPredicate{}).
   WithMgrSetup(func (..){ .. set up manager ... }).
    WithImplementation(NewImplementation(...)).
   Complete()
```
but not sure it adds a lot a benefit, because we're almost duplicating what controller-runtime provide.